An update for TerrainUtils

Trokkin opened the pull request.

Cokemonkey11 commented:
Looks pretty good. Maybe squash?

Frotty commented on wurst/util/TerrainUtils.wurst:
```diff
@@ -1,24 +1,58 @@
+
+/** represents a single terrain tile */
+public tuple tile(int id)
+let tilesX = R2I(boundMax.x - boundMin.x) div 128 + 1
```
extension funcs? @Trokkin will u update this soonish?

Trokkin commented:
So, @Frotty, it is updated though not tested because I have no possibility at the moment.
Though as always I do say that it should work.

Frotty commented:
Build fails, so I doubt that

Trokkin commented:
Err, idk why my ide didn't show me that errors
Hey, I don't like that I can't add pretty much any package due to cyclic dependencies. I assume that getterrainZ should be moved into Vector package.

Trokkin commented:
Moved .getTerrainZ() into Vectors package just to be sure that cycles won't happen.
I resolved all the errors, optimized the code not to have repeating functions and corrected usage in different packages.

Frotty commented on wurst/util/TerrainUtils.wurst:
```diff
@@ -1,91 +1,153 @@
+public function tile.getY() returns real
+	return (this.id div TILES_X) * 128. + boundMin.y
+
+// vec2 convertion
```
typo

Trokkin commented:
Tile framework completely depends on MapBounds package, since it requires minimum and maximum of the map in order to index all tiles in the map, so, until MapBounds can be somehow tested from compiletime, neither can Tile.
Yeah, it's a big PR, but it's all about one package, and I don't agree that it should be divided. What you thought is like creating a PR for each commit to work-in-progress feature. And if every next PR will depend on previous, this will slow down development speed much more.
Well, getTerrainZ() move could be in other PR, if only it wouldn't be caused by a cycle appeared when I pulled updates from master.

Frotty commented:
> Tile framework completely depends on MapBounds package, since it requires minimum and maximum of the map in order to index all tiles in the map, so, while MapBounds cannot be somehow tested from compiletime, neither can Tile.

This Test now runs successfully https://github.com/wurstscript/WurstStdlib2/commit/80110a350295910673a2aaf0b45aca0c93c84e0a#diff-9215e1d6aaac66c35bfa08182de283b1R63
If you need anything else let me know

> What you thought is like creating a PR for each commit to work-in-progress feature

No, not work in progress, but system implementation can be decoupled from actual usage, as well as unrelated extension functions and asset paths. The assets and extension functions in this PR I would insta merge. But since all this other stuff is also here, I can't.

Trokkin commented:
Er, that update to mapbounds seems like a workaround to me. Well, nevermind.

Ok, I got your idea.
Anyway, is there anything else left to do here aside from unit tests?

Frotty commented on wurst/util/TerrainUtils.wurst:
```diff
@@ -1,91 +1,149 @@
+
+constant	MAX_RANGE_SQ 	= 10.*10.
+constant 	DUMMY_ITEM_ID 	= 'wolg'
+let	   		dItem   		= CreateItem(DUMMY_ITEM_ID, 0, 0)..setVisible(false)
```
createItem(DUMMY_ITEM_ID, ZERO2)

Frotty commented on wurst/util/TerrainUtils.wurst:
```diff
@@ -1,91 +1,149 @@
+constant	MAX_RANGE_SQ 	= 10.*10.
+constant 	DUMMY_ITEM_ID 	= 'wolg'
+let	   		dItem   		= CreateItem(DUMMY_ITEM_ID, 0, 0)..setVisible(false)
+let	   		find   			= Rect(0., 0., 128., 128.)
```
these constants should adhere to naming convention. Change should be fine since they are not public.

Frotty commented:
> Er, that update to mapbounds seems like a workaround to me. Well, nevermind.

Supplying mock values at runtime is somewhat a workaround, yes. At some point we might be able to read in those values from wurst.build, but for now this shall suffice.
As long as the tests verify everything works as you expect it to on that mock data, it's already a big win.

Other than that and a couple comments I think this is ready.

Frotty commented:
so it's done?

Trokkin commented:
Yes, it is,...
...aside from extraneous tests for wc3's specific terrain functions that do not exist in wurst compiletime equivalent.

Trokkin commented:
So, @Frotty, will you merge?

Frotty commented:
I was away this week. Will take a look in the next few days.

Cokemonkey11 commented on wurst/util/TerrainUtils.wurst:
```diff
@@ -67,17 +105,39 @@ public function isTerrainLand(real x, real y) returns boolean
+public function vec2.isTerrainPlatform() returns boolean
+	return isTerrainPlatform(this)
 
 public function isTerrainWalkable(real x, real y) returns boolean
```
Source?

Cokemonkey11 commented:
LGTM

Frotty commented:
@Cokemonkey11 can you stop making single comments again? Is it so hard to click the button 50 pixels to the right?

Frotty commented:
sry forgot to submit a pending review.

Trokkin commented on wurst/_handles/primitives/Real.wurst:
```diff
@@ -21,6 +21,21 @@ public function real.sign() returns int
 public function real.round() returns int
 	return this > 0 ? (this+.5).toInt() : (this-.5).toInt()
 
+/** Rounds the input real to the nearest 'scale' divisible */
```
Because it properly divides by any `real`. E.g. (3.9).round(1,5) == (4.5)

Frotty replied:
Yeah, one usually expects round/floor functions to expect ints, not reals.
In any case, I'm not sure if the functions in general are a good fit. If I were to search for such functionality, I wouldn't search for "round". When I see `.round(412)` I'm not immediately understanding what exactly the parameter means. And since it's a simple one-liner it's usually implemented in-place anyway.
Do you even make use of roundUp/Down with scale yourself? Do you have any examples of this in other libraries?

Also, divisible might be a term, but I haven't seen it used in this context. Something like "Returns the nearest multiple of 'scale' from the input number" strikes me with much more understanding.

Trokkin commented on wurst/math/Vectors.wurst:
```diff
@@ -81,6 +81,21 @@ public function vec2.rotate(angle angl) returns vec2
 	
+let tempLoc = Location(0.,0.)
+public function vec2.getTerrainZ() returns real
+	MoveLocation(tempLoc, this.x, this.y)
```
eh? there are no extension funcs for locations in the entire stdlib, and I won't add anything else to this already huge PR

Frotty replied:
- [x] next PR

Trokkin commented on wurst/math/Maths.wurst:
```diff
@@ -157,4 +157,3 @@ public function min(real a, real b, real c, real d, real e, real f) returns real
 	
 function test_hypot_help(real x, real y)
 	hypot(x,y).assertEquals(SquareRoot(x*x + y*y), 0.0001)
-	
```
wtf, there's a newline if I edit this both online and in my IDE

Frotty replied:
then reset the Maths.wurst file..

Frotty commented on wurst/util/TerrainUtils.wurst:
```diff
@@ -1,91 +1,187 @@
+	tile(boundMax.x, boundMin.y).id.assertEquals(TILES_X - 1)
+	tile(boundMin.x, boundMax.y).id.assertEquals(TILES_X * (TILES_Y - 1))
+
+@test public function testUpperBound()
```
don't make test functions public, otherwise they will show up in autocomplete

Trokkin commented:
It's ok to merge now.

Frotty commented:
you didn't even reply to comment

Frotty commented:
🍬 finally
